interfaces/builtin/pwm_test.go

```diff
@@ -0,0 +1,182 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin_test
+
+import (
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/builtin"
+	"github.com/snapcore/snapd/interfaces/systemd"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/snaptest"
+	"github.com/snapcore/snapd/testutil"
+)
+
+type PwmInterfaceSuite struct {
+	iface                           interfaces.Interface
+	gadgetPwmSlotInfo               *snap.SlotInfo
+	gadgetPwmSlot                   *interfaces.ConnectedSlot
+	gadgetMissingNumberSlotInfo     *snap.SlotInfo
+	gadgetMissingNumberSlot         *interfaces.ConnectedSlot
+	gadgetBadNumberSlotInfo         *snap.SlotInfo
+	gadgetBadNumberSlot             *interfaces.ConnectedSlot
+	gadgetMissingChipNumberSlotInfo *snap.SlotInfo
+	gadgetMissingChipNumberSlot     *interfaces.ConnectedSlot
+	gadgetBadChipNumberSlotInfo     *snap.SlotInfo
+	gadgetBadChipNumberSlot         *interfaces.ConnectedSlot
+	gadgetBadInterfaceSlotInfo      *snap.SlotInfo
+	gadgetBadInterfaceSlot          *interfaces.ConnectedSlot
+	gadgetPlugInfo                  *snap.PlugInfo
+	gadgetPlug                      *interfaces.ConnectedPlug
+	gadgetBadInterfacePlugInfo      *snap.PlugInfo
+	gadgetBadInterfacePlug          *interfaces.ConnectedPlug
+	osPwmSlotInfo                   *snap.SlotInfo
+	osPwmSlot                       *interfaces.ConnectedSlot
+}
+
+var _ = Suite(&PwmInterfaceSuite{
+	iface: builtin.MustInterface("pwm"),
+})
+
+func (s *PwmInterfaceSuite) SetUpTest(c *C) {
+	gadgetInfo := snaptest.MockInfo(c, `
+name: my-device
+version: 0
+type: gadget
+slots:
+    my-pin:
+        interface: pwm
+        chip-number: 10
+        number: 100
+    missing-number:
+        interface: pwm
+        chip-number: 10
+    bad-number:
+        interface: pwm
+        chip-number: 10
+        number: forty-two
+    missing-chip-number:
+        interface: pwm
+        number: 100
+    bad-chip-number:
+        interface: pwm
+        chip-number: forty-two
+        number: 100
+    bad-interface-slot: other-interface
+plugs:
+    plug: pwm
+    bad-interface-plug: other-interface
```

Comment:
so in order to test the apparmor snippets I mention below, you will need to add an app to the gadget, though a hook would work too and is perhaps more realistic, so something like:
```suggestion
    bad-interface-plug: other-interface
apps:
    svc:
        command: bin/foo.sh
```

then the snippet I recommended about the AppArmor snippet test would become:

```go
func (s *PwmInterfaceSuite) TestApparmorConnectedSlot(c *C) {
	spec := &apparmor.Specification{}
	err := spec.AddConnectedPlug(s.iface, s.gadgetPlug, s.gadgetPwmSlot)
	c.Assert(err, IsNil)
	c.Assert(spec.SnippetForTag("snap.my-device.svc"), testutil.Contains, `/sys/class/pwm/pwmchip10/pwm100/* rwk`)
}
```

Reply:
@woodrow-shen ^

Reply:
Still go the error like
```
pwm_test.go:164:
    c.Assert(spec.SnippetForTag("snap.my-device.svc"), testutil.Contains, `/sys/class/pwm/pwmchip10/pwm100/* rwk`)
... container string = "" 
... elem string = "/sys/class/pwm/pwmchip10/pwm100/* rwk" 

FAIL: pwm_test.go:160: PwmInterfaceSuite.TestApparmorConnectedSlot
```

Btw, I suspect that the function name should be `TestApparmorConnectedPlug` instead of `TestApparmorConnectedSlot`? Coz I think `TestApparmorConnectedPlug` in the test code could map to `AppArmorConnectedPlug` to make sense?

And, I also found that gpio_test.go didn't have `TestApparmorConnectedPlug` as it also implemented `AppArmorConnectedPlug`, but maybe some stories behind the history...

Reply:
ok, sorry I forgot you need to mock the evalSymlink function in order for the policy to generate, something like this should work for you: https://pastebin.ubuntu.com/p/btsYXQGrGY/

Reply:
Thanks @anonymouse64 helping a lot on the testing! Just updated accordingly. TIL!